db/write_batch.cc

```diff
@@ -574,64 +605,165 @@ Status WriteBatchInternal::Iterate(const WriteBatch* wb,
 
     switch (tag) {
       case kTypeColumnFamilyValue:
-      case kTypeValue:
+      case kTypeValue: {
         assert(wb->content_flags_.load(std::memory_order_relaxed) &
                (ContentFlags::DEFERRED | ContentFlags::HAS_PUT));
-        s = handler->PutCF(column_family, key, value);
+
+        if (wb->protected_) {
+          KvProtectionInfo kv_prot_info = wb->kv_prot_infos_[found];
+          // `ValueType` is used differently in `WriteBatch` and
+          // `WriteBatch::Handler`. Here in `WriteBatch`, different `ValueType`s
+          // are used to distinguish whether an entry specifies a column family.
+          // Later in `WriteBatch::Handler`, that distinction is dropped and a
+          // single `ValueType` is used for both cases. So, this is our last
+          // chance to verify against the distinguished `ValueType`.
+          KvProtectionInfo expected_kv_prot_info;
+          expected_kv_prot_info.SetValueType(static_cast<ValueType>(tag));
+          s = kv_prot_info.VerifyAgainst(expected_kv_prot_info);
```

Comment:
It appears that we anyway would verify in MemTable::Add(). Do we need this extra one?

Reply:
Good question, it is a pretty subtle/questionable case this could catch. Imagine value type got flipped from `kTypeColumnFamilyValue` to `kTypeValue` or vice-versa. The check in `MemTable::Add()` wouldn't catch it because `MemTableInserter::PutCF()` uses `kTypeValue` in either case.

I think such a flip wouldn't actually matter since our checks for the other components would catch it (it would likely cause CF ID to be wrong, and if not, key would definitely be wrong due to start reading it from the wrong offset). Should we remove checks for flips that don't matter? I'm not sure.

Reply:
Done, removed it by dropping the CF vs non-CF distinction (e.g., `kTypeColumnFamilyValue` vs `kTypeValue`) before adding to the protection info. We just use the non-CF variant everywhere.